File: semantic_sweep/parser.py
"""Parse exported Power BI TMDL semantic models into normalized ``ModelCard`` objects.

A semantic model is a folder ``<name>.SemanticModel`` produced by ``fab export`` with a
``definition/`` tree (``tables/*.tmdl``, ``relationships.tmdl``, ``expressions.tmdl``). We only
extract what the similarity scorer needs: tables, columns (+ type + hidden), measures (+ DAX),
relationships, and the logical/physical data source.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

_USAGE_METRICS_RE = re.compile(r"usage\s*metrics", re.IGNORECASE)
_SQL_DATABASE_RE = re.compile(r'Sql\.Database\(\s*"([^"]+)"\s*,\s*"([^"]+)"', re.IGNORECASE)
_ENTITY_RE = re.compile(r"^\s*entityName:\s*(.+?)\s*$")
_SCHEMA_RE = re.compile(r"^\s*schemaName:\s*(.+?)\s*$")


def _tab_indent(line: str) -> int:
    return len(line) - len(line.lstrip("\t"))


def _unquote(name: str) -> str:
    name = name.strip()
    if name.startswith("'") and name.endswith("'") and len(name) >= 2:
        name = name[1:-1]
    return name.strip()


@dataclass
class Measure:
    """A single DAX measure (name + raw expression)."""

    name: str
    dax: str


@dataclass
class Column:
    """A table column with its data type and hidden flag."""

    table: str
    name: str
    data_type: str | None = None
    hidden: bool = False

    @property
    def qualified(self) -> str:
        """Return the ``Table[Column]`` qualified name (lower-cased)."""
        return f"{self.table}[{self.name}]".lower()


@dataclass
class ModelCard:  # pylint: disable=too-many-instance-attributes
    """Normalized, comparable view of one semantic model."""

    name: str
    workspace: str
    path: Path
    tables: list[str] = field(default_factory=list)
    columns: list[Column] = field(default_factory=list)
    measures: list[Measure] = field(default_factory=list)
    relationships: list[tuple[str, str]] = field(default_factory=list)
    source_logical: set[tuple[str, str]] = field(default_factory=set)
    source_physical: set[tuple[str, str]] = field(default_factory=set)
    has_rls: bool = False
    has_calc_groups: bool = False
    system_generated: bool = False

    @property
    def measure_count(self) -> int:
        """Number of measures in the model."""
        return len(self.measures)

    @property
    def table_count(self) -> int:
        """Number of tables in the model."""
        return len(self.tables)

    @property
    def qualified_columns(self) -> set[str]:
        """Set of lower-cased ``table[column]`` names."""
        return {c.qualified for c in self.columns}


def _read_measure(lines: list[str], i: int) -> tuple[Measure, int]:
    """Read one measure starting at line ``i``; return it and the next line index."""
    line = lines[i]
    indent = _tab_indent(line)
    match = re.match(r"^\t+measure\s+(.+?)\s*=\s*(.*)$", line)
    parts: list[str] = []
    if match is None:
        # A blank/placeholder measure declared with no `= expression` on the line (e.g. a
        # freshly-added `measure Medida` that carries only child properties). Real estates contain
        # these, so treat it as an empty measure instead of crashing the whole scan.
        name = _unquote(re.sub(r"^\t+measure\s+", "", line).strip())
        i += 1
    else:
        name = _unquote(match.group(1))
        inline = match.group(2).strip()
        i += 1
        if inline == "```":  # triple-backtick fenced multi-line expression
            while i < len(lines) and lines[i].strip() != "```":
                parts.append(lines[i].strip())
                i += 1
            i += 1  # consume closing fence
        elif inline == "":  # `=` on its own line: DAX continues at a deeper indent
            # The exporter emits a leading whitespace-only line before the body, so a blank line is
            # part of the expression block, not its terminator. Stop only at the first non-blank line
            # shallow enough to be a child property (indent + 1) or a sibling/parent (<= indent).
            while i < len(lines):
                cur = lines[i]
                if cur.strip() == "":
                    i += 1
                    continue
                if _tab_indent(cur) <= indent + 1:
                    break
                parts.append(cur.strip())
                i += 1
        else:  # single-line measure
            parts.append(inline)
    # Skip the measure's child property lines (formatString, displayFolder, kpi, ...).
    while i < len(lines):
        cur = lines[i]
        if cur.strip() == "":
            i += 1
            continue
        if _tab_indent(cur) <= indent:
            break
        i += 1
    return Measure(name=name, dax=" ".join(p for p in parts if p)), i


def _read_column(lines: list[str], i: int, table: str) -> tuple[Column, int]:
    """Read one column starting at line ``i``; return it and the next line index."""
    indent = _tab_indent(lines[i])
    body = lines[i].strip()[len("column") :].strip()
    if body.startswith("'"):
        name = body[1 : body.index("'", 1)]
    else:
        name = body.split("=")[0].strip()
    data_type: str | None = None
    hidden = False
    i += 1
    while i < len(lines):
        cur = lines[i]
        if cur.strip() == "":
            i += 1
            continue
        if _tab_indent(cur) <= indent:
            break
        stripped = cur.strip()
        if stripped.startswith("dataType:"):
            data_type = stripped.split(":", 1)[1].strip()
        elif stripped == "isHidden" or stripped.startswith("isHidden:"):
            hidden = True
        i += 1
    return Column(table=table, name=_unquote(name), data_type=data_type, hidden=hidden), i


def _parse_table_file(path: Path) -> tuple[str | None, list[Column], list[Measure], bool]:
    """Parse one ``tables/*.tmdl`` file into (table_name, columns, measures, is_calc_group)."""
    lines = path.read_text(encoding="utf-8").splitlines()
    table_name: str | None = None
    columns: list[Column] = []
    measures: list[Measure] = []
    is_calc_group = False
    i = 0
    while i < len(lines):
        line = lines[i]
        stripped = line.strip()
        if _tab_indent(line) == 0 and stripped.startswith("table "):
            table_name = _unquote(stripped[len("table ") :])
        elif stripped.startswith("calculationGroup"):
            is_calc_group = True
        elif re.match(r"^\t+measure\s", line):
            measure, i = _read_measure(lines, i)
            measures.append(measure)
            continue
        elif re.match(r"^\t+column\s", line):
            column, i = _read_column(lines, i, table_name or "")
            columns.append(column)
            continue
        i += 1
    return table_name, columns, measures, is_calc_group


def _parse_relationships(path: Path) -> list[tuple[str, str]]:
    """Parse ``relationships.tmdl`` into normalized (fromColumn, toColumn) lower-cased pairs."""
    if not path.exists():
        return []
    rels: list[tuple[str, str]] = []
    from_col: str | None = None
    for line in path.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if stripped.startswith("fromColumn:"):
            from_col = stripped.split(":", 1)[1].strip().lower()
        elif stripped.startswith("toColumn:") and from_col is not None:
            rels.append((from_col, stripped.split(":", 1)[1].strip().lower()))
            from_col = None
    return rels


def _parse_logical_source(table_files: list[Path]) -> set[tuple[str, str]]:
    """Collect (schema, entity) tuples from Direct Lake / entity partitions across table files."""
    entities: set[tuple[str, str]] = set()
    for path in table_files:
        schema: str | None = None
        entity: str | None = None
        for line in path.read_text(encoding="utf-8").splitlines():
            if (match := _SCHEMA_RE.match(line)) is not None:
                schema = _unquote(match.group(1)).lower()
            elif (match := _ENTITY_RE.match(line)) is not None:
                entity = _unquote(match.group(1)).lower()
        if entity is not None:
            entities.add((schema or "", entity))
    return entities


def _parse_physical_source(definition_dir: Path) -> set[tuple[str, str]]:
    """Collect (endpoint, database) tuples from any ``Sql.Database(...)`` in the definition."""
    physical: set[tuple[str, str]] = set()
    for path in definition_dir.rglob("*.tmdl"):
        for endpoint, database in _SQL_DATABASE_RE.findall(path.read_text(encoding="utf-8")):
            physical.add((endpoint.lower(), database.lower()))
    return physical


def parse_model(model_dir: Path, workspace: str) -> ModelCard:
    """Parse a single ``<name>.SemanticModel`` folder into a :class:`ModelCard`."""
    definition = model_dir / "definition"
    name = model_dir.name.removesuffix(".SemanticModel")
    card = ModelCard(name=name, workspace=workspace, path=model_dir)

    tables_dir = definition / "tables"
    table_files = sorted(tables_dir.glob("*.tmdl")) if tables_dir.exists() else []
    for table_file in table_files:
        table_name, columns, measures, is_calc_group = _parse_table_file(table_file)
        if table_name is None:
            continue
        card.tables.append(table_name)
        card.columns.extend(columns)
        card.measures.extend(measures)
        card.has_calc_groups = card.has_calc_groups or is_calc_group

    card.relationships = _parse_relationships(definition / "relationships.tmdl")
    card.source_logical = _parse_logical_source(table_files)
    card.source_physical = _parse_physical_source(definition) if definition.exists() else set()
    card.has_rls = _has_rls(definition)
    card.system_generated = bool(_USAGE_METRICS_RE.search(name))
    return card


def _has_rls(definition_dir: Path) -> bool:
    """Return True if any role defines a table permission (row-level security)."""
    if not definition_dir.exists():
        return False
    for path in definition_dir.rglob("*.tmdl"):
        if "tablePermission" in path.read_text(encoding="utf-8"):
            return True
    return False


def load_models(models_root: Path, *, keep_empty: bool = False) -> list[ModelCard]:
    """Load every semantic model under ``models_root`` (``<workspace>/<name>.SemanticModel``).

    Models with no tables (default lakehouse/warehouse/staging artifacts) are skipped unless
    ``keep_empty`` is set.
    """
    cards: list[ModelCard] = []
    for model_dir in sorted(models_root.glob("*/*.SemanticModel")):
        card = parse_model(model_dir, workspace=model_dir.parent.name)
        if not card.tables and not keep_empty:
            continue
        cards.append(card)
    return cards

File: engine/parser.ts
// TMDL parsing in the browser (or Node). Ported from semantic_sweep/parser.py.
// Consumes in-memory files: { path, text }[] — environment-agnostic.

import type { Column, Measure, ModelCard } from "./types";

export interface InputFile {
  path: string;
  text: string;
}

const USAGE_METRICS_RE = /usage\s*metrics/i;
// One connection arg: a literal string value, OR a bare identifier (an M query parameter / let-bound
// name, e.g. `Sql.Database(ServerParam, DatabaseParam)`) captured symbolically so parameterized
// connections still contribute comparable physical-source evidence instead of being silently dropped.
const ARG_SRC = String.raw`(?:"([^"]+)"|(\w+))`;
// `\b` matters here: without it "Sql.Database(" also matches inside "PostgreSQL.Database(" and
// "MySQL.Database(" (both connector names literally end in "...SQL"), miscategorizing those sources
// as SQL Server. PostgreSQL and MySQL get their own dedicated (and correctly `\b`-anchored) regexes
// below so this fix doesn't silently drop coverage that previously worked only by substring accident.
const SQL_DATABASE_RE = new RegExp(String.raw`\bSql\.Database\(\s*` + ARG_SRC + String.raw`\s*,\s*` + ARG_SRC, "gi");
const POSTGRESQL_RE = new RegExp(
  String.raw`\bPostgreSQL\.Database\(\s*` + ARG_SRC + String.raw`\s*,\s*` + ARG_SRC,
  "gi",
);
const MYSQL_RE = new RegExp(String.raw`\bMySQL\.Database\(\s*` + ARG_SRC + String.raw`\s*,\s*` + ARG_SRC, "gi");
const SNOWFLAKE_RE = new RegExp(
  String.raw`\bSnowflake\.Databases?\(\s*` + ARG_SRC + String.raw`\s*,\s*` + ARG_SRC,
  "gi",
);
const DATABRICKS_RE = new RegExp(
  String.raw`\bDatabricks\.Catalogs\(\s*` + ARG_SRC + String.raw`\s*,\s*` + ARG_SRC,
  "gi",
);
// BigQuery rarely carries a comparable 2-arg identity (often zero args, or an options record); capture
// a single literal/identifier arg (billing project) when present, else there's nothing to fingerprint.
const BIGQUERY_RE = new RegExp(String.raw`\bGoogleBigQuery\.Database\(\s*` + ARG_SRC + "?", "gi");
// File-based connectors: the file path is the physical identity. Only a simple literal path or a
// single bare parameter reference is captured; a concatenation expression (e.g. folder & name &
// ".csv") is dynamic per-row/per-parameter and cannot be fingerprinted statically.
const CSV_RE = new RegExp(String.raw`\bCsv\.Document\(\s*File\.Contents\(\s*` + ARG_SRC + String.raw`\s*\)`, "gi");
const PARQUET_RE = new RegExp(
  String.raw`\bParquet\.Document\(\s*File\.Contents\(\s*` + ARG_SRC + String.raw`\s*\)`,
  "gi",
);
// Bare M parameter names common enough in templated connection setups (e.g. every environment's model
// has a query parameter literally called "Server"/"Database") that a shared PARAMETER NAME alone (not
// an actual literal value) must not manufacture physical-source evidence between unrelated models.
const GENERIC_PARAM_NAMES = new Set([
  "server", "database", "host", "hostname", "warehouse", "instance", "catalog", "schema",
  "port", "endpoint", "source", "db", "path", "filepath", "file", "url", "connection", "conn",
]);
// Composite / DirectQuery-to-Power-BI-dataset reference: a model built ON another semantic model.
const ANALYSIS_SERVICES_RE = /AnalysisServices\.Databases?\(\s*"([^"]*)"(?:\s*,\s*"([^"]+)")?/gi;
// Fabric/PBI DirectQuery-to-dataset via the dedicated connector (e.g. PowerBIDatasets / PowerPlatform.Dataflows-style).
const PBI_DATASETS_RE = /PowerBIDatasets\s*\(/i;
// M navigation step `{[Name="DatasetName"]}[Data]` — how the upstream dataset name appears when it
// isn't an inline argument to AnalysisServices.Database(s).
const NAV_NAME_RE = /\{\s*\[\s*Name\s*=\s*"([^"]+)"\s*\]\s*\}\s*\[\s*Data\s*\]/gi;
const SCHEMA_RE = /^\s*schemaName:\s*(.+?)\s*$/;
const ENTITY_RE = /^\s*entityName:\s*(.+?)\s*$/;

const tabIndent = (line: string): number => line.length - line.replace(/^\t+/, "").length;

function unquote(name: string): string {
  let n = name.trim();
  if (n.startsWith("'") && n.endsWith("'") && n.length >= 2) n = n.slice(1, -1);
  return n.trim();
}

const splitLines = (text: string): string[] => text.split(/\r\n|\r|\n/);

function readMeasure(lines: string[], start: number): [Measure, number] {
  let i = start;
  const line = lines[i];
  const indent = tabIndent(line);
  const m = /^\t+measure\s+(.+?)\s*=\s*(.*)$/.exec(line);
  const name = unquote(m ? m[1] : "");
  const inline = (m ? m[2] : "").trim();
  const parts: string[] = [];
  i += 1;
  if (inline === "```") {
    while (i < lines.length && lines[i].trim() !== "```") {
      parts.push(lines[i].trim());
      i += 1;
    }
    i += 1;
  } else if (inline === "") {
    // The exporter emits a leading whitespace-only line before the body, so a blank line is part of
    // the expression block, not its terminator. Stop only at the first non-blank line shallow enough
    // to be a child property (indent + 1) or a sibling/parent (<= indent).
    while (i < lines.length) {
      const cur = lines[i];
      if (cur.trim() === "") {
        i += 1;
        continue;
      }
      if (tabIndent(cur) <= indent + 1) break;
      parts.push(cur.trim());
      i += 1;
    }
  } else {
    parts.push(inline);
  }
  while (i < lines.length) {
    const cur = lines[i];
    if (cur.trim() === "") {
      i += 1;
      continue;
    }
    if (tabIndent(cur) <= indent) break;
    i += 1;
  }
  return [{ name, dax: parts.filter((p) => p).join(" ") }, i];
}

function readColumn(lines: string[], start: number, table: string): [Column, number] {
  let i = start;
  const indent = tabIndent(lines[i]);
  const body = lines[i].trim().slice("column".length).trim();
  let name: string;
  if (body.startsWith("'")) name = body.slice(1, body.indexOf("'", 1));
  else name = body.split("=")[0].trim();
  let dataType: string | null = null;
  let hidden = false;
  i += 1;
  while (i < lines.length) {
    const cur = lines[i];
    if (cur.trim() === "") {
      i += 1;
      continue;
    }
    if (tabIndent(cur) <= indent) break;
    const stripped = cur.trim();
    if (stripped.startsWith("dataType:")) dataType = stripped.split(/:(.*)/s)[1].trim();
    else if (stripped === "isHidden" || stripped.startsWith("isHidden:")) hidden = true;
    i += 1;
  }
  return [{ table, name: unquote(name), dataType, hidden }, i];
}

interface TableParse {
  tableName: string | null;
  columns: Column[];
  measures: Measure[];
  isCalcGroup: boolean;
}

function parseTableFile(text: string): TableParse {
  const lines = splitLines(text);
  let tableName: string | null = null;
  const columns: Column[] = [];
  const measures: Measure[] = [];
  let isCalcGroup = false;
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    const stripped = line.trim();
    if (tabIndent(line) === 0 && stripped.startsWith("table ")) {
      tableName = unquote(stripped.slice("table ".length));
    } else if (stripped.startsWith("calculationGroup")) {
      isCalcGroup = true;
    } else if (/^\t+measure\s/.test(line)) {
      const [measure, next] = readMeasure(lines, i);
      measures.push(measure);
      i = next;
      continue;
    } else if (/^\t+column\s/.test(line)) {
      const [column, next] = readColumn(lines, i, tableName ?? "");
      columns.push(column);
      i = next;
      continue;
    }
    i += 1;
  }
  return { tableName, columns, measures, isCalcGroup };
}

function parseRelationships(text: string): string[] {
  const rels: string[] = [];
  let fromCol: string | null = null;
  for (const line of splitLines(text)) {
    const stripped = line.trim();
    if (stripped.startsWith("fromColumn:")) {
      fromCol = stripped.split(/:(.*)/s)[1].trim().toLowerCase();
    } else if (stripped.startsWith("toColumn:") && fromCol !== null) {
      rels.push(`${fromCol}\u0000${stripped.split(/:(.*)/s)[1].trim().toLowerCase()}`);
      fromCol = null;
    }
  }
  return rels;
}

function logicalFromTable(text: string): string | null {
  let schema: string | null = null;
  let entity: string | null = null;
  for (const line of splitLines(text)) {
    const s = SCHEMA_RE.exec(line);
    if (s) {
      schema = unquote(s[1]).toLowerCase();
      continue;
    }
    const e = ENTITY_RE.exec(line);
    if (e) entity = unquote(e[1]).toLowerCase();
  }
  return entity !== null ? `${schema ?? ""}\u0000${entity}` : null;
}

const normPath = (p: string): string => p.replace(/\\/g, "/");

interface ModelFiles {
  workspace: string;
  name: string;
  files: Array<{ rel: string; text: string }>;
}

function groupByModel(files: InputFile[]): ModelFiles[] {
  const groups = new Map<string, ModelFiles>();
  for (const f of files) {
    const parts = normPath(f.path).split("/");
    const idx = parts.findIndex((p) => p.endsWith(".SemanticModel"));
    if (idx === -1 || idx === 0) continue;
    const workspace = parts[idx - 1];
    const name = parts[idx].replace(/\.SemanticModel$/, "");
    const rel = parts.slice(idx + 1).join("/");
    const key = `${workspace}/${name}`;
    if (!groups.has(key)) groups.set(key, { workspace, name, files: [] });
    groups.get(key)!.files.push({ rel, text: f.text });
  }
  return [...groups.values()];
}

// Add a "server\u0000database"-shaped entry for every match of a 2-positional-arg connector regex
// (Sql.Database, PostgreSQL.Database, Snowflake.Databases, Databricks.Catalogs all share this shape).
function addTwoArgPhysicalSource(card: ModelCard, re: RegExp, text: string): void {
  for (const m of text.matchAll(re)) {
    const arg1 = m[1] ?? m[2];
    const arg2 = m[3] ?? m[4];
    if (!arg1 || !arg2) continue;
    const bothParams = m[2] !== undefined && m[4] !== undefined;
    if (bothParams && GENERIC_PARAM_NAMES.has(arg1.toLowerCase()) && GENERIC_PARAM_NAMES.has(arg2.toLowerCase())) {
      continue; // templated parameter names alone are not comparable evidence
    }
    card.sourcePhysical.add(`${arg1.toLowerCase()}\u0000${arg2.toLowerCase()}`);
  }
}

// Add a "value\u0000"-shaped entry (empty second component, matching the scanner's `db ?? ""`
// convention) for a single-arg connector regex (BigQuery billing project, CSV/Parquet file path).
function addSingleArgPhysicalSource(card: ModelCard, re: RegExp, text: string): void {
  for (const m of text.matchAll(re)) {
    const arg = m[1] ?? m[2];
    if (!arg) continue;
    if (m[2] !== undefined && GENERIC_PARAM_NAMES.has(arg.toLowerCase())) continue;
    card.sourcePhysical.add(`${arg.toLowerCase()}\u0000`);
  }
}

function parseModel(mf: ModelFiles): ModelCard {
  const card: ModelCard = {
    name: mf.name,
    workspace: mf.workspace,
    tables: [],
    columns: [],
    measures: [],
    relationships: [],
    sourceLogical: new Set(),
    sourcePhysical: new Set(),
    hasRls: false,
    hasCalcGroups: false,
    systemGenerated: USAGE_METRICS_RE.test(mf.name),
  };
  const defFiles = mf.files.filter((f) => f.rel.startsWith("definition/"));
  const tableFiles = defFiles
    .filter((f) => /^definition\/tables\/.+\.tmdl$/.test(f.rel))
    .sort((a, b) => a.rel.localeCompare(b.rel));

  for (const tf of tableFiles) {
    const parsed = parseTableFile(tf.text);
    if (parsed.tableName === null) continue;
    card.tables.push(parsed.tableName);
    card.columns.push(...parsed.columns);
    card.measures.push(...parsed.measures);
    card.hasCalcGroups = card.hasCalcGroups || parsed.isCalcGroup;
    const logical = logicalFromTable(tf.text);
    if (logical) card.sourceLogical.add(logical);
  }

  const rels = defFiles.find((f) => f.rel === "definition/relationships.tmdl");
  if (rels) card.relationships = parseRelationships(rels.text);

  const derived = new Set<string>();
  let compositeSeen = false;
  for (const f of defFiles) {
    if (!f.rel.endsWith(".tmdl")) continue;
    addTwoArgPhysicalSource(card, SQL_DATABASE_RE, f.text);
    addTwoArgPhysicalSource(card, POSTGRESQL_RE, f.text);
    addTwoArgPhysicalSource(card, MYSQL_RE, f.text);
    addTwoArgPhysicalSource(card, SNOWFLAKE_RE, f.text);
    addTwoArgPhysicalSource(card, DATABRICKS_RE, f.text);
    addSingleArgPhysicalSource(card, BIGQUERY_RE, f.text);
    addSingleArgPhysicalSource(card, CSV_RE, f.text);
    addSingleArgPhysicalSource(card, PARQUET_RE, f.text);
    // Collect M navigation names once per file; the deepest (last) nav is the queried dataset.
    const navNames = [...f.text.matchAll(NAV_NAME_RE)].map((n) => n[1].trim());
    let fileHasComposite = false;
    for (const m of f.text.matchAll(ANALYSIS_SERVICES_RE)) {
      // A PBI/AS dataset endpoint => this is a composite model chained onto another dataset.
      if (/pbiazure|powerbi|analysis\.windows\.net/i.test(m[1])) {
        compositeSeen = true;
        fileHasComposite = true;
        if (m[2]) derived.add(m[2].trim());
      }
    }
    if (PBI_DATASETS_RE.test(f.text)) {
      compositeSeen = true;
      fileHasComposite = true;
    }
    // If a composite source in this file didn't carry an inline dataset name, recover it from the
    // last `{[Name="…"]}[Data]` navigation step (the dataset the model actually DirectQueries).
    if (fileHasComposite && navNames.length) derived.add(navNames[navNames.length - 1]);
    if (f.text.includes("tablePermission")) card.hasRls = true;
  }
  if (compositeSeen && derived.size === 0) derived.add("a Power BI dataset");
  if (derived.size) card.derivedFrom = [...derived];
  return card;
}

export function loadModelsFromFiles(files: InputFile[], keepEmpty = false): ModelCard[] {
  const cards = groupByModel(files)
    .map(parseModel)
    .filter((c) => keepEmpty || c.tables.length > 0);
  cards.sort((a, b) => `${a.workspace}/${a.name}`.localeCompare(`${b.workspace}/${b.name}`));
  return cards;
}
